cmd/occollector/app/builder/pipelines_builder.go

```diff
@@ -0,0 +1,164 @@
+	pipelineProcessors := make(PipelineProcessors)
+
+	for _, pipeline := range eb.config.Pipelines {
+		firstProcessor, err := eb.buildPipeline(pipeline)
+		if err != nil {
+			return nil, err
+		}
+		pipelineProcessors[pipeline] = firstProcessor
+	}
+
+	return pipelineProcessors, nil
+}
+
+// Builds a pipeline of processors. Returns the first processor in the pipeline.
+// The last processor in the pipeline will be plugged to fan out the data into exporters
+// that are configured for this pipeline.
+func (eb *PipelinesBuilder) buildPipeline(
+	pipelineCfg *configmodels.Pipeline,
+) (*builtProcessor, error) {
+
+	// Build the pipeline backwards.
+
+	// First create a consumer junction point that fans out the data to all exporters.
+	var tc consumer.TraceConsumer
+	var mc consumer.MetricsConsumer
+
+	switch pipelineCfg.InputType {
+	case configmodels.TracesDataType:
+		tc = eb.buildFanoutExportersTraceConsumer(pipelineCfg.Exporters)
+	case configmodels.MetricsDataType:
+		mc = eb.buildFanoutExportersMetricsConsumer(pipelineCfg.Exporters)
+	}
+
+	// Now build the processors backwards, starting from the last one.
+	// The last processor points to consumer which fans out to exporters, then
+	// the processor itself becomes a consumer for the one that precedes it in
+	// in the pipeline and so on.
+	for i := len(pipelineCfg.Processors) - 1; i >= 0; i-- {
+		procName := pipelineCfg.Processors[i]
+		procCfg := eb.config.Processors[procName]
+
+		factory := factories.GetProcessorFactory(procCfg.Type())
+
+		// This processor must point to the next consumer and then
+		// it becomes the next for the previous one (previous in the pipeline,
+		// which we will build in the next loop iteration).
+		var err error
+		switch pipelineCfg.InputType {
+		case configmodels.TracesDataType:
+			tc, err = factory.CreateTraceProcessor(tc, procCfg)
+		case configmodels.MetricsDataType:
+			mc, err = factory.CreateMetricsProcessor(mc, procCfg)
+		}
+
+		if err != nil {
+			return nil, fmt.Errorf("error creating processor %q in pipeline %q: %v",
+				procName, pipelineCfg.Name, err)
+		}
+	}
+
+	return &builtProcessor{tc, mc}, nil
+}
+
+// Converts the list of exporter names to a list of corresponding builtExporters.
+func (eb *PipelinesBuilder) getBuiltExportersByNames(exporterNames []string) []*builtExporter {
+	var result []*builtExporter
+	for _, name := range exporterNames {
+		exporter := eb.exporters[eb.config.Exporters[name]]
+		result = append(result, exporter)
+	}
+
+	return result
+}
+
+func (eb *PipelinesBuilder) buildFanoutExportersTraceConsumer(exporterNames []string) consumer.TraceConsumer {
+	builtExporters := eb.getBuiltExportersByNames(exporterNames)
+
+	var exporterConsumers []consumer.TraceConsumer
+	for _, impl := range builtExporters {
+		exporterConsumers = append(exporterConsumers, impl.tc)
+	}
+
+	// Optimize for the case when there is only one exporter, no need to create junction point.
```

Comment:
nit: in practice this is not going to make a real difference since it is done once on initialization but you could instead check `len(builtExporters)` and return `builtExporters[0].tc` without having to allocate the slice. You can also ue this length to create exporterConsumers with the exact size that it needs.

Reply:
same for the metrics function